### PR TITLE
Upgrade to Django 4.1.x

### PR DESCRIPTION
**Pull request by theskumar**
- Upgrade Django, wagtail-purge, Django-pwned-password,django-basic-auth-ip-whitelist
- Move logic from DeleteView -> delete() to form_valid()
- Make test more robust
- Upgrade dj-database-url to 2.x, and enable CONN_HEALTH_CHECKS
- Bump django-two-factor-auth to 1.15.x that has confirmed support for Django 4.1
- Bump django-ratelimit to 4.1.0
- django-pwned-passwords seems be abandoned https://github.com/jamiecounsell/django-pwned-passwords/pull/11 (replaced with it with it's fork)

### CODE REVIEW

**Comment by theskumar** [comment]
Update: Rebased on `main` and use official release version of `wagtail-purge=0.3.0`

**Comment by theskumar** [comment]
@frjo Update with main and resolved conflicts.

**Comment by frjo** [comment]
Test are failing, found this regarding `SuccessURLAllowedHostsMixin`:

> * The undocumented ``django.contrib.auth.views.SuccessURLAllowedHostsMixin``
>   mixin is replaced by ``RedirectURLMixin``.

https://github.com/django/django/blob/21382e2743d06efbf5623e7c9b6dccf2a325669b/docs/releases/4.1.txt#L573-L574

**Comment by theskumar** [comment]
Updated the code. The logic behind the SuccessURLAllowedHostsMixin is was already covered in the get_redirect_url ~mixin~ function.

**Review by frjo — APPROVED** [review]

**Comment by theskumar** [comment]
- Upgrade dj-database-url to 2.x, and enable CONN_HEALTH_CHECKS
- Bump django-two-factor-auth to 1.15.x that has confirmed support for Django 4.1
- Bump django-ratelimit to 4.1.0

**Comment by frjo** [comment]
@theskumar I merge in a couple of you other update PRs so this one has a small merge conflict now.

Will run this on test for a some days to be sure but it seems good to go in my testing.

**Comment by theskumar** [comment]
Thanks for the quick merging. I've resolve the conflicts with `main`.

**Comment by theskumar** [comment]
@frjo Would you be more comfortable if I stack the wagtail upgrade #3627  PR onto this django 4x PR? from testing perspective?

**Comment by frjo** [comment]
@theskumar Yes, I think it would be good to do the Wagtail 5 on top of Django 4. I think we can get this PR quickly tested and then merged in to main.

For Django 4.2.x LTS we need Wagtail 5 I believe. But that where I would like to go before the end of the year.

**Comment by theskumar** [comment]
Done. Updated the base of the wagtail 5x to this PR.

File: requirements.txt [hunk 1]
```diff
@@ -23,17 +23,18 @@ django-heroku==0.3.1
...
 django-select2==8.0.0
 django-slack==5.18.0
 django-storages==1.13.2
 django-tables2==2.5.1
 django-tinymce==3.5.0
-django-two-factor-auth==1.14.0
+django-two-factor-auth==1.15.5
 django-web-components==0.1.1
-django==3.2.22
+django==4.1.12
```
Comment (frjo) [inline requirements.txt]: Django 4.1.13 was just released.

**Pull request merged**